### PR TITLE
Momentum scheduler

### PR DESCRIPTION
**Pull request by ZwwWayne**
There are some cases that the momentum needs to be scheduled, for example, scheduling the momentum together with Cyclic LR could accelerate the model convergence.
This PR adds a momentum scheduler and a cyclic LR scheduler, and I have been using these two features in my project for three months and they work fine.

### CODE REVIEW

**Comment by hellock** [comment]
MomentumUpdaterHook and LrUpdaterHook have highly overlapped codes. We may consider merging them.

File: mmcv/runner/hooks/lr_updater.py [hunk 1]
```diff
@@ -201,8 +201,9 @@ def get_lr(self, runner, base_lr):
 @HOOKS.register_module
 class CosineLrUpdaterHook(LrUpdaterHook):
 
-    def __init__(self, target_lr=0, **kwargs):
+    def __init__(self, target_lr=0, as_ratio=False, **kwargs):
         self.target_lr = target_lr
```
Comment (xvjiarui) [inline mmcv/runner/hooks/lr_updater.py:205]: `target_lr` is confusing when `as_ratio=True`

File: mmcv/runner/hooks/lr_updater.py [hunk 2]
```diff
@@ -212,5 +213,70 @@ def get_lr(self, runner, base_lr):
...
+                progress = curr_iter - start_iter
+                return annealing_cos(base_lr * start_ratio,
+                                     base_lr * end_ratio,
+                                     progress / (end_iter - start_iter))
+
+
+def annealing_cos(start, end, pct):
+    'Cosine anneal from `start` to `end` as pct goes from 0.0 to 1.0.'
+    cos_out = cos(pi * pct) + 1
+    return end + 0.5 * (start - end) * cos_out
```
Comment (xvjiarui) [inline mmcv/runner/hooks/lr_updater.py:282]: Should we apply it to `CosineLrUpdater` as well?

File: mmcv/runner/runner.py [hunk 3]
```diff
@@ -254,7 +269,7 @@ def train(self, data_loader, **kwargs):
         self.model.train()
         self.mode = 'train'
         self.data_loader = data_loader
-        self._max_iters = self._max_epochs * len(data_loader)
+        # self._max_iters = self._max_epochs * len(data_loader)
```
Comment (xvjiarui) [inline mmcv/runner/runner.py:272]: We may delete this comment.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,28 +1,28 @@
 exclude: ^tests/data/
 repos:
-- repo: https://gitlab.com/pycqa/flake8
```
Comment (yhcao6) [inline .pre-commit-config.yaml:3]: No need to modify this file.

File: mmcv/runner/hooks/lr_updater.py [hunk 5]
```diff
@@ -212,5 +213,70 @@ def get_lr(self, runner, base_lr):
...
             max_progress = runner.max_iters
-        return self.target_lr + 0.5 * (base_lr - self.target_lr) * \
-            (1 + cos(pi * (progress / max_progress)))
+        if self.as_ratio:
+            target_lr = base_lr * self.target_lr
+            return target_lr + 0.5 * (base_lr - target_lr) * \
+                (1 + cos(pi * (progress / max_progress)))
+        else:
+            return self.target_lr + 0.5 * (base_lr - self.target_lr) * \
+                (1 + cos(pi * (progress / max_progress)))
```
Comment (yhcao6) [inline mmcv/runner/hooks/lr_updater.py:222]: It can merge into one statement.
```
        if self.as_ratio:
            target_lr = base_lr * self.target_lr
        return target_lr + 0.5 * (base_lr - target_lr) * \
               (1 + cos(pi * (progress / max_progress)))
```

File: mmcv/runner/hooks/lr_updater.py [hunk 6]
```diff
@@ -212,5 +213,70 @@ def get_lr(self, runner, base_lr):
...
+                 target_ratio=[10, 1e-4],
+                 cyclic_times=1,
+                 step_ratio_up=0.4,
+                 **kwargs):
+        if isinstance(target_ratio, float):
+            target_ratio = [target_ratio, target_ratio / 1e5]
+        elif isinstance(target_ratio, list):
+            target_ratio = (
+                target_ratio + target_ratio[0] / 1e5
+                if len(target_ratio) == 1 else target_ratio)
```
Comment (yhcao6) [inline mmcv/runner/hooks/lr_updater.py:238]: If `len(target_ratio) == 1`, then `target_ratio = target_ratio + target_ratio[0] / 1e5`.
However, `target_ratio` is a list, `target_ratio[0]` is a float number, I guess this will report an error.

File: mmcv/runner/hooks/lr_updater.py [hunk 7]
```diff
@@ -212,5 +213,70 @@ def get_lr(self, runner, base_lr):
...
+            curr_iter %= max_iter_per_phase
+            if start_iter <= curr_iter < end_iter:
+                progress = curr_iter - start_iter
+                return annealing_cos(base_lr * start_ratio,
+                                     base_lr * end_ratio,
+                                     progress / (end_iter - start_iter))
+
+
+def annealing_cos(start, end, pct):
+    'Cosine anneal from `start` to `end` as pct goes from 0.0 to 1.0.'
```
Comment (yhcao6) [inline mmcv/runner/hooks/lr_updater.py:280]: Use triple double-quoted strings should be used for docstring.

Also explain args and return.

File: mmcv/runner/hooks/lr_updater.py (hunk 6 above)
Comment (yhcao6) [inline mmcv/runner/hooks/lr_updater.py:238]: No need parentheses.

File: mmcv/runner/hooks/lr_updater.py [hunk 8]
```diff
@@ -212,5 +213,70 @@ def get_lr(self, runner, base_lr):
...
+        assert len(target_ratio) == 2, \
+            '"target_ratio" must be list of two floats'
+        assert 0 <= step_ratio_up < 1.0, \
+            '"step_ratio_up" must be in range [0,1)'
+
+        self.target_ratio = target_ratio
+        self.cyclic_times = cyclic_times
+        self.step_ratio_up = step_ratio_up
+        self.lr_phases = []  # init lr_phases
+        # currently only support by_epoch=False
```
Comment (yhcao6) [inline mmcv/runner/hooks/lr_updater.py:249]: This comment is redundant.

File: .pre-commit-config.yaml (hunk 4 above)
Comment (ZwwWayne) [reply .pre-commit-config.yaml:3]: Sometimes this path cannot be correctly processed by pre-commit.

File: mmcv/runner/hooks/logger/tensorboard.py [hunk 9]
```diff
@@ -52,6 +52,10 @@ def log(self, runner):
             else:
                 self.writer.add_scalar(tag, runner.log_buffer.output[var],
                                        runner.iter)
+        self.writer.add_scalar('learning_rate',
+                               runner.current_lr()[0], runner.iter)
+        self.writer.add_scalar('momentum',
+                               runner.current_momentum()[0], runner.iter)
```
Comment (hellock) [inline mmcv/runner/hooks/logger/tensorboard.py:58]: Other logger hooks such as PaviLoggerHook may share the same modification.

File: mmcv/runner/hooks/logger/text.py [hunk 10]
```diff
@@ -100,6 +103,15 @@ def log(self, runner):
...
             log_dict['time'] = runner.log_buffer.output['time']
             log_dict['data_time'] = runner.log_buffer.output['data_time']
+
+            print_momentum = False
+            for hook in runner._hooks:
+                if isinstance(hook, MomentumUpdaterHook):
+                    print_momentum = True
+            # only record momentum of the first param group in training
+            if print_momentum:
+                log_dict['momentum'] = runner.current_momentum()[0]
```
Comment (hellock) [inline mmcv/runner/hooks/logger/text.py:113]: Looks a little tricky.

File: mmcv/runner/hooks/lr_updater.py [hunk 11]
```diff
@@ -201,8 +201,9 @@ def get_lr(self, runner, base_lr):
 @HOOKS.register_module
 class CosineLrUpdaterHook(LrUpdaterHook):
 
-    def __init__(self, target_lr=0, **kwargs):
-        self.target_lr = target_lr
+    def __init__(self, target=0, as_ratio=False, **kwargs):
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py:204]: - We may rename the hook to `CosineAnnealingLrUpdaterHook`?
- Rename `target_lr` to `min_lr` to keep it consistent with PyTorch APIs.
- `min_lr=None, min_lr_ratio=None` instead of `target=0, as_ratio=False`
- We may add an argument `last_epoch=-1`.

File: mmcv/runner/hooks/lr_updater.py [hunk 12]
```diff
@@ -212,5 +213,67 @@ def get_lr(self, runner, base_lr):
...
-        return self.target_lr + 0.5 * (base_lr - self.target_lr) * \
-            (1 + cos(pi * (progress / max_progress)))
+        if self.as_ratio:
+            target_lr = base_lr * self.target
+        else:
+            target_lr = self.target
+        return annealing_cos(base_lr, target_lr, progress / max_progress)
+
+
+class CyclicLrUpdaterHook(LrUpdaterHook):
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py:223]: Add a docstring.

File: mmcv/runner/hooks/momentum_updater.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
+from __future__ import division
```
Comment (hellock) [inline mmcv/runner/hooks/momentum_updater.py:1]: Remove this line.

File: mmcv/runner/hooks/lr_updater.py [hunk 14]
```diff
@@ -199,11 +199,13 @@ def get_lr(self, runner, base_lr):
...
-class CosineLrUpdaterHook(LrUpdaterHook):
+class CosineAnealingLrUpdaterHook(LrUpdaterHook):
 
-    def __init__(self, target_lr=0, **kwargs):
-        self.target_lr = target_lr
-        super(CosineLrUpdaterHook, self).__init__(**kwargs)
+    def __init__(self, min_lr=None, min_lr_ratio=None, **kwargs):
+        assert (min_lr is None) ^ (min_lr_ratio is None)
+        self.min_lr = min_lr
+        self.min_lr_ratio = min_lr_ratio
```
Comment (hellock) [inline mmcv/runner/hooks/lr_updater.py:207]: Assert that these two arguments cannot be set at the same time.
